finality-aleph/src/sync/forest/mod.rs

```diff
@@ -0,0 +1,721 @@
+use std::collections::{
+    hash_map::{Entry, OccupiedEntry, VacantEntry},
+    HashMap, HashSet,
+};
+
+use crate::sync::{BlockIdentifier, Header, Justification, PeerId};
+
+mod vertex;
+
+use vertex::{JustificationAddResult, Vertex};
+
+type BlockIdFor<J> = <<J as Justification>::Header as Header>::Identifier;
+
+pub struct JustificationWithParent<J: Justification> {
+    pub justification: J,
+    pub parent: BlockIdFor<J>,
+}
+
+enum VertexHandle<'a, I: PeerId, J: Justification> {
+    HopelessFork,
+    BelowMinimal,
+    HighestFinalized,
+    Unknown(VacantEntry<'a, BlockIdFor<J>, VertexWithChildren<I, J>>),
+    Candidate(OccupiedEntry<'a, BlockIdFor<J>, VertexWithChildren<I, J>>),
+}
+
+/// Our interest in a block referred to by a vertex, including the information about whom we expect to have the block.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum Interest<I: PeerId> {
+    /// We are not interested in this block.
+    Uninterested,
+    /// We would like to have this block.
+    Required(HashSet<I>),
+    /// We would like to have this block and its the highest on its branch.
+    TopRequired(HashSet<I>),
+}
+
+/// What can go wrong when inserting data into the forest.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum Error {
+    HeaderMissingParentId,
+    IncorrectParentState,
+    IncorrectVertexState,
+    ParentNotImported,
+    TooNew,
+}
+
+pub struct VertexWithChildren<I: PeerId, J: Justification> {
+    vertex: Vertex<I, J>,
+    children: HashSet<BlockIdFor<J>>,
+}
+
+impl<I: PeerId, J: Justification> VertexWithChildren<I, J> {
+    fn new() -> Self {
+        Self {
+            vertex: Vertex::new(),
+            children: HashSet::new(),
+        }
+    }
+
+    fn add_child(&mut self, child: BlockIdFor<J>) {
+        self.children.insert(child);
+    }
+}
+
+// How deep can the forest be, vaguely based on two sessions ahead, which is the most we expect to
+// ever need worst case scenario.
+const MAX_DEPTH: u32 = 1800;
+
+pub struct Forest<I: PeerId, J: Justification> {
+    vertices: HashMap<BlockIdFor<J>, VertexWithChildren<I, J>>,
+    top_required: HashSet<BlockIdFor<J>>,
+    root_id: BlockIdFor<J>,
+    root_children: HashSet<BlockIdFor<J>>,
+    compost_bin: HashSet<BlockIdFor<J>>,
+}
+
+impl<I: PeerId, J: Justification> Forest<I, J> {
+    pub fn new(highest_justified: BlockIdFor<J>) -> Self {
+        Self {
+            vertices: HashMap::new(),
+            top_required: HashSet::new(),
+            root_id: highest_justified,
+            root_children: HashSet::new(),
+            compost_bin: HashSet::new(),
+        }
+    }
+
+    fn get_mut(&mut self, id: &BlockIdFor<J>) -> VertexHandle<I, J> {
```

Comment:
Now this method can insert an entry that is `TooNew` and is used in `state` method call. We should probably add here a check for that..

Reply:
This is an internal method though. Besides, we cannot really guarantee it based on the thing we return.

Reply:
Not sure I understand. `insert_id` is also a private function and has this check.

This function is used in `state`, which is public, so calling it might create an entry we do not want to create, right? I guess adding this check to `state` might also make sense.

Reply:
Well yeah, the check has to happen somewhere and `insert_id` was the most natural place for it. And `state` does not return the actual `Entry`, so it doesn't let you mutate anything outside of the struct.

Reply:
Ah, yes, the vacant entry is not inserted unless you modify it.